## Assignment Topic:

In this project, you will build a regression model using the Keras library to model the same data about concrete compressive strength that we used in labs 3.p In this course project, you will build a regression model using the deep learning Keras library, and then you will experiment with increasing the number of training epochs and changing number of hidden layers and you will see how changing these parameters impacts the performance of the model.

### Changing Number of Hidden Layers - 1

***

## Data Preprocessing

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

(1030, 9)
So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

In [6]:
concrete_data_columns = concrete_data.columns

target = concrete_data["Strength"]
predictors = concrete_data.drop(["Strength", "Age"], axis=1)

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  
0            1040.0           676.0  
1            1055.0           676.0  
2             932.0           594.0  
3             932.0           594.0  
4             978.4           825.5

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate  
0          0.862735       -1.217079  
1          1.055651       -1.217079  
2         -0.526262       -2.239829  
3         -0.526262       -2.239829  
4          0.070492        0.647569

In [10]:
n_cols = predictors_norm.shape[1]  # number of predictors
n_cols

7

***

## Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

In [11]:
import warnings
warnings.filterwarnings("ignore")
import keras

In [12]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [13]:
# Define regression model
def regression_model():
    # 1-Create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # 2-Compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

In [16]:
model = regression_model()

In [17]:
epochs = 100
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/100
23/23 [==============================] - 0s 2ms/step - loss: 24729.6836
Epoch 2/100
23/23 [==============================] - 0s 2ms/step - loss: 7458.7744
Epoch 3/100
23/23 [==============================] - 0s 2ms/step - loss: 4838.7363
Epoch 4/100
23/23 [==============================] - 0s 817us/step - loss: 3816.5142
Epoch 5/100
23/23 [==============================] - 0s 923us/step - loss: 3126.7258
Epoch 6/100
23/23 [==============================] - 0s 970us/step - loss: 2653.5710
Epoch 7/100
23/23 [==============================] - 0s 954us/step - loss: 2261.7776
Epoch 8/100
23/23 [==============================] - 0s 928us/step - loss: 1919.7834
Epoch 9/100
23/23 [==============================] - 0s 931us/step - loss: 1639.4254
Epoch 10/100
23/23 [==============================] - 0s 1ms/step - loss: 1402.8625
Epoch 11/100
23/23 [==============================] - 0s 949us/step - loss: 1215.5603
Epoch 12/100
23/23 [==============================] - ETA: 0s - loss: 

23/23 [==============================] - 0s 2ms/step - loss: 169.9814
Epoch 97/100
23/23 [==============================] - 0s 2ms/step - loss: 163.3132
Epoch 98/100
23/23 [==============================] - 0s 1ms/step - loss: 166.8344
Epoch 99/100
23/23 [==============================] - 0s 798us/step - loss: 163.6809
Epoch 100/100
23/23 [==============================] - 0s 810us/step - loss: 163.7814


In [18]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 886us/step - loss: 171.0525


171.05252075195312

***

### Mean Squared Error

In [19]:
from sklearn.metrics import mean_squared_error

In [20]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(f"Mean: {mean}\nStandard Deviation: {standard_deviation}")

Mean: 171.05250175678216
Standard Deviation: 0.0


In [21]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("Mean Squared Error " + str(i + 1) +": "+ str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print("\n" + str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " + str(epochs) + "\n")
print("Mean: "+ str(mean))
print("Standard Deviation: "+ str(standard_deviation))

Mean Squared Error 1: 141.35183715820312
Mean Squared Error 2: 160.56797790527344
Mean Squared Error 3: 161.73928833007812
Mean Squared Error 4: 187.150390625
Mean Squared Error 5: 179.1959686279297
Mean Squared Error 6: 193.71246337890625
Mean Squared Error 7: 176.01434326171875
Mean Squared Error 8: 154.51031494140625
Mean Squared Error 9: 163.10130310058594
Mean Squared Error 10: 166.17807006835938
Mean Squared Error 11: 161.30532836914062
Mean Squared Error 12: 144.8494110107422
Mean Squared Error 13: 146.93914794921875
Mean Squared Error 14: 165.2831268310547
Mean Squared Error 15: 148.51202392578125
Mean Squared Error 16: 137.09300231933594
Mean Squared Error 17: 150.0919189453125
Mean Squared Error 18: 129.07554626464844
Mean Squared Error 19: 139.4480743408203
Mean Squared Error 20: 164.4240264892578
Mean Squared Error 21: 137.41229248046875
Mean Squared Error 22: 153.1256103515625
Mean Squared Error 23: 170.66787719726562
Mean Squared Error 24: 146.36798095703125
Mean Squared 